### DISCUSSION TITLE
Next song/s functionality broken in check_next

### DISCUSSION THREAD

**Discussion by MattMski** [savonet/liquidsoap#2993] [User Support]
**Describe the bug**
A clear and concise description of what the bug is. Add some script and log extract so we can see what is happening. Don't forget to add some context although not necessarily your whole script..

Hey  guys, so had to downgrade back to liquidsoap 1.4.4 after finding these issues that don't allow me to do older functionality that I use. check_next appears to be broken, when used with a playlist, and outputs the current song when testing. 

**To Reproduce**
A minimal script to reproduce the issue is the most efficient way to have your bug addressed and fixed very quickly!

**Expected behavior**
Check_next should output the next song.

**Version details**

- OS: [e.g. Debian, OSX] Debian 11
- Version [e.g. `1.3.4`] 2.1.2

**Install method**
Did you install via `opam`? Using your distribution's packages? From source? Opam 

**Common issues**
Can't decode or encode in some audio format? Check if you have [installed the correct dependencies](https://www.liquidsoap.info/#download)

**Comment by toots** [comment]
Hi @MattMski

Are you able to describe your issue more in details?

`check_next` is present and tested here: https://github.com/savonet/liquidsoap/blob/main/tests/streams/197.liq so, unless you can explain more how you use it, what you expect it do and what's it's doing, it's a little hard to guess what issue you are having.

Thanks!

**Comment by MattMski** [comment]
Hey @toots sure.

In my code, I use check_next to grab the next song from a playlist of music. I then take that, add it to meta as the next song, and send it to my shoutcast server to also display as the next song.

When I tested the code with Liquidsoap 2, my current song was coming in as the next song.

I also did update map_metadata to the renamed function in Liquidsoap 2.

``` 
main = playlist("/music/main", check_next=update_next)

#Update Next Song
def update_next(m)
  n = request.metadata(m)
  next:= n["artist"] ^ " - " ^ n["title"]
  true
end

#Update Meta
def update_meta(m)  
	if m["source"] == "main" then
	[("next","#{!next}")]
	end	
end

#Update the metadata with values like nextsong everytime it sees new metadata
radio = map_metadata(update_meta,radio) 
```

**Comment by toots** [comment]
I think that you should add some debugging to verify which script is called with which values. Something like this:
```ruby
#Update Next Song
def update_next(m)
  n = request.metadata(m)
  next:= m["artist"] ^ " - " ^ m["title"]
  print("updating next song to: #{!next}")
  true
end

main = playlist("/music/main", check_next=update_next)

#Update Meta
def update_meta(m)  
        print("update_meta called with #{m} and {!next}")
	if m["source"] == "main" then
	[("next","#{!next}")]
	end	
end

#Update the metadata with values like nextsong everytime it sees new metadata
radio = map_metadata(update_meta,radio) 
```

**Comment by MattMski** [comment]
Hey @toots, sure I'll try to test it out again in the next few days and let you know what the results are.

**Comment by MattMski** [comment]
Hey @toots, been a while but finally looked into it. Thanks for your example. I tested the print statement and it is indeed getting the next song. However, I am using metadata.map later on and that is not catching the updated next song. 

So both the current and next song end up the same value. It seems like check_next runs after update_meta all the time and this is breaking the code.

```
# Update Next Song
def update_next(m)
  n = request.metadata(m)
  next := n["artist"] ^ " - " ^ n["title"]
  print("updating next song to: #{!next}")
  true
end

# Update Meta
def update_meta(m)
	title = ref(m["title"])
	print("updating meta to current:  #{!title}")
	print("updating meta to next: #{!next}")
	if m["title"] == "Advert:" or m["source"] == "request" or m["source"] == "DJ" then
	[("next","")]
	elsif m["source"] == "main" then
	[("next","#{!next}")]
	else
	[("next","")]
	end
end

main = playlist("/music/main", check_next=update_next)
radio = metadata.map(update_meta, radio)
```

logs:
updating meta to current:  Advert:
updating meta to next:
updating next song to: REO Speedwagon - Time for Me to Fly
updating meta to current:  Time for Me to Fly
updating meta to next:

**Comment by toots** [comment]
As discussed, I believe that you should use an array and queue multiple tracks in advance. Something like this:
```liquidsoap
next_tracks = ref([])

# Update Next Song
def update_next(m)
  n = request.metadata(m)
  
  next_track = n["artist"] ^ n["title"]
  next := [...!next, next_track]
  print("updating next song to: #{!next}")
  
  true
end

# Update Meta
def update_meta(m)
	title = ref(m["title"])
	print("updating meta to current:  #{!title}")
	print("updating meta to next: #{!next}")
	if m["title"] == "Advert:" or m["source"] == "request" or m["source"] == "DJ" then
	  [("next","")]
	elsif m["source"] == "main" then
	  let [next_track, ...remaining_tracks] = !next
          next := remaining_tracks
	  [("next", next_track)]
	else
	  [("next","")]
	end
end

main = playlist("/music/main", check_next=update_next)
radio = metadata.map(update_meta, radio)
```
